ntp-proto/src/packet/error.rs

```diff
@@ -1,24 +1,54 @@
 use std::fmt::Display;
 
+use crate::NtpPacket;
+
 #[derive(Debug)]
-pub enum PacketParsingError {
+pub enum ParsingError<T> {
     InvalidVersion(u8),
     IncorrectLength,
     MalformedNtsExtensionFields,
     MalformedNonce,
-    DecryptError,
+    DecryptError(T),
+}
+
+impl<T> ParsingError<T> {
+    pub(super) fn coerce<U>(self) -> Option<ParsingError<U>> {
+        match self {
+            ParsingError::InvalidVersion(v) => Some(ParsingError::InvalidVersion(v)),
+            ParsingError::IncorrectLength => Some(ParsingError::IncorrectLength),
+            ParsingError::MalformedNtsExtensionFields => {
+                Some(ParsingError::MalformedNtsExtensionFields)
+            }
+            ParsingError::MalformedNonce => Some(ParsingError::MalformedNonce),
+            ParsingError::DecryptError(_) => None,
+        }
+    }
 }
 
-impl Display for PacketParsingError {
+impl ParsingError<std::convert::Infallible> {
+    pub(super) fn force<U>(self) -> ParsingError<U> {
+        match self {
+            ParsingError::InvalidVersion(v) => ParsingError::InvalidVersion(v),
+            ParsingError::IncorrectLength => ParsingError::IncorrectLength,
+            ParsingError::MalformedNtsExtensionFields => ParsingError::MalformedNtsExtensionFields,
+            ParsingError::MalformedNonce => ParsingError::MalformedNonce,
+            ParsingError::DecryptError(_) => unreachable!(),
+        }
+    }
+}
+
+pub type PacketParsingError<'a> = ParsingError<NtpPacket<'a>>;
+
+impl<T> Display for ParsingError<T> {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
             Self::InvalidVersion(version) => f.write_fmt(format_args!("Invalid version {version}")),
             Self::IncorrectLength => f.write_str("Incorrect packet length"),
             Self::MalformedNtsExtensionFields => f.write_str("Malformed nts extension fields"),
             Self::MalformedNonce => f.write_str("Malformed nonce (likely invalid length)"),
-            Self::DecryptError => f.write_str("Failed to decrypt NTS extension fields"),
+            Self::DecryptError(_) => f.write_str("Failed to decrypt NTS extension fields"),
         }
     }
 }
```

Comment:
I want to suggest

```rust
impl<T> ParsingError<T> {
    pub(super) fn get_decrypt_error<U>(self) -> Result<T, ParsingError<U>> {
        match self {
            ParsingError::InvalidVersion(v) => Err(ParsingError::InvalidVersion(v)),
            ParsingError::IncorrectLength => Err(ParsingError::IncorrectLength),
            ParsingError::MalformedNtsExtensionFields => {
                Err(ParsingError::MalformedNtsExtensionFields)
            }
            ParsingError::MalformedNonce => Err(ParsingError::MalformedNonce),
            ParsingError::DecryptError(decrypt_error) => Ok(decrypt_error),
        }
    }
}

impl ParsingError<std::convert::Infallible> {
    pub(super) fn generalize<U>(self) -> ParsingError<U> {
        match self {
            ParsingError::InvalidVersion(v) => ParsingError::InvalidVersion(v),
            ParsingError::IncorrectLength => ParsingError::IncorrectLength,
            ParsingError::MalformedNtsExtensionFields => ParsingError::MalformedNtsExtensionFields,
            ParsingError::MalformedNonce => ParsingError::MalformedNonce,
            ParsingError::DecryptError(decrypt_error) => match decrypt_error {},
        }
    }
}
```

- this removes an `unwrap` below and would generally improve control flow in a few places
- uses exhaustive matching on `Infallibe` instead of an `unreachable!` (probably would compile away? not sure how rust deals with empty sum types)
- a function `forall U. Infallible -> U`  "generalizes the type". I think that name fits better than "force"

Reply:
Implemented.